graphics/src/AssimpLoader.cc

```diff
@@ -0,0 +1,733 @@
+  for (unsigned childIdx = 0; childIdx < _node->mNumChildren; ++childIdx)
+  {
+    auto child_node = _node->mChildren[childIdx];
+    // Finally recursive call to explore subnode
+    this->RecursiveStoreBoneNames(_scene, child_node, _boneNames);
+  }
+}
+
+//////////////////////////////////////////////////
+void AssimpLoader::Implementation::RecursiveSkeletonCreate(const aiNode* _node,
+    SkeletonNode* _parent, const math::Matrix4d& _transform,
+    const std::unordered_set<std::string> &_boneNames) const
+{
+  if (_node == nullptr || _parent == nullptr)
+    return;
+  // First explore this node
+  auto nodeName = ToString(_node->mName);
+  auto boneExist = _boneNames.find(nodeName) != _boneNames.end();
+  auto nodeTrans = this->ConvertTransform(_node->mTransformation);
+  auto skelNode = _parent;
+
+  if (boneExist)
+  {
+    skelNode = new SkeletonNode(
+        _parent, nodeName, nodeName, SkeletonNode::JOINT);
+    skelNode->SetTransform(nodeTrans);
+  }
+
+  nodeTrans = _transform * nodeTrans;
+
+  for (unsigned childIdx = 0; childIdx < _node->mNumChildren; ++childIdx)
+  {
+    this->RecursiveSkeletonCreate(
+        _node->mChildren[childIdx], skelNode, nodeTrans, _boneNames);
+  }
+}
+
+//////////////////////////////////////////////////
+MaterialPtr AssimpLoader::Implementation::CreateMaterial(
+    const aiScene* _scene, unsigned _matIdx, const std::string& _path) const
+{
+  MaterialPtr mat = std::make_shared<Material>();
+  aiColor4D color;
+  auto& assimpMat = _scene->mMaterials[_matIdx];
+  auto ret = assimpMat->Get(AI_MATKEY_COLOR_DIFFUSE, color);
+  if (ret == AI_SUCCESS)
+  {
+    mat->SetDiffuse(this->ConvertColor(color));
+  }
+  ret = assimpMat->Get(AI_MATKEY_COLOR_AMBIENT, color);
+  if (ret == AI_SUCCESS)
+  {
+    mat->SetAmbient(this->ConvertColor(color));
+  }
+  ret = assimpMat->Get(AI_MATKEY_COLOR_SPECULAR, color);
+  if (ret == AI_SUCCESS)
+  {
+    mat->SetSpecular(this->ConvertColor(color));
+  }
+  ret = assimpMat->Get(AI_MATKEY_COLOR_EMISSIVE, color);
+  if (ret == AI_SUCCESS)
+  {
+    mat->SetEmissive(this->ConvertColor(color));
+  }
+  float shininess;
+  ret = assimpMat->Get(AI_MATKEY_SHININESS, shininess);
+  if (ret == AI_SUCCESS)
+  {
+    mat->SetShininess(shininess);
+  }
+  float opacity = 1.0;
+  ret = assimpMat->Get(AI_MATKEY_OPACITY, opacity);
+  mat->SetTransparency(1.0 - opacity);
+  mat->SetBlendFactors(opacity, 1.0 - opacity);
+  // TODO(luca) more than one texture, Gazebo assumes UV index 0
+  Pbr pbr;
+  aiString texturePath(_path.c_str());
+  ret = assimpMat->GetTexture(aiTextureType_DIFFUSE, 0, &texturePath);
+  // TODO(luca) check other arguments,
+  // type of mappings to be UV, uv index, blend mode
+  if (ret == AI_SUCCESS)
+  {
+    // Check if the texture is embedded or not
+    auto [texName, texData] = this->LoadTexture(_scene,
+        texturePath, this->GenerateTextureName(_scene, assimpMat, "Diffuse"));
+    if (texData != nullptr)
+      mat->SetTextureImage(texName, texData);
```

Comment:
Hi  @luca-della-vedova,
Do you know how we can combine multiple diffuse in here? I want to implement it but my knowledge is weak.